File: src/main/ipc/dbHandlers.ts
import { ipcMain } from 'electron'
import { DatabaseService } from '../services/database'

export function registerDbHandlers(db: DatabaseService) {
  // Books
  ipcMain.handle('db:getBooks', () => db.getBooks())
  ipcMain.handle('db:getBook', (_e, id: number) => db.getBook(id))
  ipcMain.handle('db:deleteBook', (_e, id: number) => db.deleteBook(id))

  // Reading Progress
  ipcMain.handle('db:getProgress', (_e, bookId: number) => db.getReadingProgress(bookId))
  ipcMain.handle('db:updateProgress', (_e, bookId: number, progress: any) => {
    db.upsertReadingProgress(bookId, progress)
    db.updateBookLastOpened(bookId)
  })

  // Bookmarks
  ipcMain.handle('db:getBookmarks', (_e, bookId: number) => db.getBookmarks(bookId))
  ipcMain.handle('db:addBookmark', (_e, bookmark: any) => db.addBookmark(bookmark))
  ipcMain.handle('db:deleteBookmark', (_e, id: number) => db.deleteBookmark(id))
  ipcMain.handle('db:updateBookmarkTitle', (_e, id: number, title: string) => {
    console.log('IPC updateBookmarkTitle called:', { id, title })
    const result = db.updateBookmarkTitle(id, title)
    console.log('IPC updateBookmarkTitle result:', result)
    return result
  })

  // Highlights
  ipcMain.handle('db:getHighlights', (_e, bookId: number) => db.getHighlights(bookId))
  ipcMain.handle('db:addHighlight', (_e, highlight: any) => db.addHighlight(highlight))
  ipcMain.handle('db:deleteHighlight', (_e, id: number) => db.deleteHighlight(id))

  // Notes
  ipcMain.handle('db:getNotes', (_e, bookId: number) => db.getNotes(bookId))
  ipcMain.handle('db:addNote', (_e, note: any) => db.addNote(note))
  ipcMain.handle('db:updateNote', (_e, id: number, content: string) => db.updateNote(id, content))
  ipcMain.handle('db:deleteNote', (_e, id: number) => db.deleteNote(id))

  // Bookshelves
  ipcMain.handle('db:getBookshelves', () => db.getBookshelves())
  ipcMain.handle('db:addBookshelf', (_e, name: string) => db.addBookshelf(name))
  ipcMain.handle('db:deleteBookshelf', (_e, id: number) => db.deleteBookshelf(id))
  ipcMain.handle('db:renameBookshelf', (_e, id: number, name: string) => db.renameBookshelf(id, name))
  ipcMain.handle('db:getBooksInShelf', (_e, shelfId: number) => db.getBooksInShelf(shelfId))
  ipcMain.handle('db:addBookToShelf', (_e, shelfId: number, bookId: number) => db.addBookToShelf(shelfId, bookId))
  ipcMain.handle('db:removeBookFromShelf', (_e, shelfId: number, bookId: number) => db.removeBookFromShelf(shelfId, bookId))
  ipcMain.handle('db:getShelvesForBook', (_e, bookId: number) => db.getShelvesForBook(bookId))

  // Reading Sessions
  ipcMain.handle('db:startReadingSession', (_e, bookId: number) => db.startReadingSession(bookId))
  ipcMain.handle('db:endReadingSession', (_e, sessionId: number) => db.endReadingSession(sessionId))
  ipcMain.handle('db:updateReadingSessionDuration', (_e, sessionId: number, durationSeconds: number) => db.updateReadingSessionDuration(sessionId, durationSeconds))
  ipcMain.handle('db:getReadingTime', (_e, bookId: number) => db.getReadingTimeForBook(bookId))
  ipcMain.handle('db:getAllReadingTime', () => db.getReadingTimeForAllBooks())
  ipcMain.handle('db:getAllReadingProgress', () => db.getAllReadingProgress())

  // Settings
  ipcMain.handle('db:getSettings', () => db.getSettings())
  ipcMain.handle('db:updateSettings', (_e, settings: any) => db.updateSettings(settings))

  // Per-book settings
  ipcMain.handle('db:getBookSettings', (_e, bookId: number) => db.getBookSettings(bookId))
  ipcMain.handle('db:updateBookSettings', (_e, bookId: number, settings: any) => db.updateBookSettings(bookId, settings))
  ipcMain.handle('db:deleteBookSettings', (_e, bookId: number) => db.deleteBookSettings(bookId))
}

File: particlebook-cpp/src/handlers/DbHandlers.cpp
#include "DbHandlers.h"
#include "BridgeServer.h"
#include "WebViewHost.h"
#include "services/DatabaseService.h"
#include <windows.h>
#include <filesystem>
#include <cstdio>
#include <ctime>

void RegisterDbHandlers(BridgeServer* bridge, DatabaseService* db)
{
    using Handler = BridgeServer::MethodHandler;

    // ── Books ───────────────────────────────────────
    bridge->RegisterMethod("db:getBooks", [db](const json&) {
        return db->GetBooks();
    });

    bridge->RegisterMethod("db:getBook", [db](const json& p) {
        auto b = db->GetBook(p["id"].get<int>());
        if (!b.is_null()) db->UpdateBookLastOpened(p["id"].get<int>());
        return b.is_null() ? json(nullptr) : b;
    });

    bridge->RegisterMethod("db:deleteBook", [db](const json& p) {
        int id = p.value("id", -1);
        if (id < 0) return json::object();
        db->DeleteBook(id);
        db->FlushSync();
        return json::object();
    });

    // ── Reading Progress ────────────────────────────
    bridge->RegisterMethod("db:updateProgress", [db](const json& p) {
        int bid = p["bookId"].get<int>();
        auto prog = p.value("progress", json::object());
        double pc = prog.value("progress", -1.0);
        int pn = prog.value("page", -1);
        auto progStr = prog.dump();
        wchar_t exePath[MAX_PATH];
        GetModuleFileNameW(nullptr, exePath, MAX_PATH);
        auto logPath = std::filesystem::path(exePath).parent_path() / "debug.log";
        FILE* lf = _wfopen(logPath.c_str(), L"a");
        if (lf) {
            time_t now = time(nullptr);
            char timeBuf[32];
            strftime(timeBuf, sizeof(timeBuf), "%H:%M:%S", localtime(&now));
            fprintf(lf, "[%s] db:updateProgress bookId=%d progress=%.2f page=%d json=%s\n",
                    timeBuf, bid, pc, pn, progStr.c_str());
            fclose(lf);
        }
        db->UpsertProgress(bid, prog);
        db->FlushSync();
        return json::object();
    });

    bridge->RegisterMethod("db:getProgress", [db](const json& p) {
        int bid = p["bookId"].get<int>();
        auto r = db->GetProgress(bid);
        auto result = r.is_null() ? json::object() : r;
        // Dump entire reading_progress array for diagnosis
        auto all = db->DumpProgress();
        wchar_t exePath[MAX_PATH];
        GetModuleFileNameW(nullptr, exePath, MAX_PATH);
        auto logPath = std::filesystem::path(exePath).parent_path() / "debug.log";
        FILE* lf = _wfopen(logPath.c_str(), L"a");
        if (lf) {
            time_t now = time(nullptr);
            char timeBuf[32];
            strftime(timeBuf, sizeof(timeBuf), "%H:%M:%S", localtime(&now));
            fprintf(lf, "[%s] db:getProgress bookId=%d result=%s all=%s\n",
                    timeBuf, bid, result.dump().c_str(), all.c_str());
            fclose(lf);
        }
        return result;
    });

    // ── Bookmarks ───────────────────────────────────
    bridge->RegisterMethod("db:getBookmarks", [db](const json& p) {
        return db->GetBookmarks(p["bookId"].get<int>());
    });
    bridge->RegisterMethod("db:addBookmark", [db](const json& p) {
        db->AddBookmark(p["bm"]);
        return json::object();
    });
    bridge->RegisterMethod("db:deleteBookmark", [db](const json& p) {
        db->DeleteBookmark(p["id"].get<int>());
        return json::object();
    });
    bridge->RegisterMethod("db:updateBookmarkTitle", [db](const json& p) {
        db->UpdateBookmarkTitle(p["id"].get<int>(), p["title"].get<std::string>());
        return json::object();
    });

    // ── Highlights ──────────────────────────────────
    bridge->RegisterMethod("db:getHighlights", [db](const json& p) {
        return db->GetHighlights(p["bookId"].get<int>());
    });
    bridge->RegisterMethod("db:addHighlight", [db](const json& p) {
        db->AddHighlight(p["hl"]);
        return json::object();
    });
    bridge->RegisterMethod("db:deleteHighlight", [db](const json& p) {
        db->DeleteHighlight(p["id"].get<int>());
        return json::object();
    });

    // ── Notes ───────────────────────────────────────
    bridge->RegisterMethod("db:getNotes", [db](const json& p) {
        return db->GetNotes(p["bookId"].get<int>());
    });
    bridge->RegisterMethod("db:addNote", [db](const json& p) {
        db->AddNote(p["note"]);
        return json::object();
    });
    bridge->RegisterMethod("db:updateNote", [db](const json& p) {
        db->UpdateNote(p["id"].get<int>(), p["content"].get<std::string>());
        return json::object();
    });
    bridge->RegisterMethod("db:deleteNote", [db](const json& p) {
        db->DeleteNote(p["id"].get<int>());
        return json::object();
    });

    // ── Settings ────────────────────────────────────
    bridge->RegisterMethod("db:getSettings", [db](const json&) {
        return db->GetSettings();
    });
    bridge->RegisterMethod("db:updateSettings", [db](const json& p) {
        db->UpdateSettings(p["settings"]);
        return json::object();
    });
    bridge->RegisterMethod("db:getBookSettings", [db](const json& p) {
        return db->GetBookSettings(p["bookId"].get<int>());
    });
    bridge->RegisterMethod("db:updateBookSettings", [db](const json& p) {
        db->UpdateBookSettings(p["bookId"].get<int>(), p["settings"]);
        return json::object();
    });
    bridge->RegisterMethod("db:deleteBookSettings", [db](const json& p) {
        db->DeleteBookSettings(p["bookId"].get<int>());
        return json::object();
    });

    // ── Bookshelves ─────────────────────────────────
    bridge->RegisterMethod("db:getBookshelves", [db](const json&) {
        return db->GetBookshelves();
    });
    bridge->RegisterMethod("db:addBookshelf", [db](const json& p) {
        return db->AddBookshelf(p["name"].get<std::string>());
    });
    bridge->RegisterMethod("db:deleteBookshelf", [db](const json& p) {
        db->DeleteBookshelf(p["id"].get<int>());
        return json::object();
    });
    bridge->RegisterMethod("db:renameBookshelf", [db](const json& p) {
        db->RenameBookshelf(p["id"].get<int>(), p["name"].get<std::string>());
        return json::object();
    });
    bridge->RegisterMethod("db:getBooksInShelf", [db](const json& p) {
        auto ids = db->GetBooksInShelf(p["shelfId"].get<int>());
        return json(ids);
    });
    bridge->RegisterMethod("db:addBookToShelf", [db](const json& p) {
        db->AddBookToShelf(p["shelfId"].get<int>(), p["bookId"].get<int>());
        return json::object();
    });
    bridge->RegisterMethod("db:removeBookFromShelf", [db](const json& p) {
        db->RemoveBookFromShelf(p["shelfId"].get<int>(), p["bookId"].get<int>());
        return json::object();
    });
    bridge->RegisterMethod("db:getShelvesForBook", [db](const json& p) {
        auto ids = db->GetShelvesForBook(p["bookId"].get<int>());
        return json(ids);
    });

    // ── Reading Sessions ────────────────────────────
    bridge->RegisterMethod("db:startReadingSession", [db](const json& p) {
        return db->StartReadingSession(p["bookId"].get<int>());
    });
    bridge->RegisterMethod("db:endReadingSession", [db](const json& p) {
        db->EndReadingSession(p["sessionId"].get<int>());
        return json::object();
    });
    bridge->RegisterMethod("db:updateReadingSessionDuration", [db](const json& p) {
        db->UpdateReadingSessionDuration(p["sessionId"].get<int>(), p["duration"].get<int>());
        return json::object();
    });
    bridge->RegisterMethod("db:getReadingTime", [db](const json& p) {
        return db->GetReadingTimeForBook(p["bookId"].get<int>());
    });
    bridge->RegisterMethod("db:getAllReadingTime", [db](const json&) {
        return db->GetAllReadingTime();
    });
    bridge->RegisterMethod("db:getAllReadingProgress", [db](const json&) {
        return db->GetAllReadingProgress();
    });

    // ── Book Sources ────────────────────────────────
    bridge->RegisterMethod("bookSource:getAll", [db](const json&) {
        return db->GetBookSources();
    });
    bridge->RegisterMethod("bookSource:get", [db](const json& p) {
        return db->GetBookSource(p["id"].get<int>());
    });
    bridge->RegisterMethod("bookSource:insert", [db](const json& p) {
        return db->InsertBookSource(p["source"]);
    });
    bridge->RegisterMethod("bookSource:update", [db](const json& p) {
        db->UpdateBookSource(p["id"].get<int>(), p["updates"]);
        return json::object();
    });
    bridge->RegisterMethod("bookSource:delete", [db](const json& p) {
        db->DeleteBookSource(p["id"].get<int>());
        return json::object();
    });
    bridge->RegisterMethod("bookSource:toggle", [db](const json& p) {
        db->ToggleBookSource(p["id"].get<int>());
        return json::object();
    });
    bridge->RegisterMethod("bookSource:clearAll", [db](const json&) {
        db->ClearAllBookSources();
        return json::object();
    });
}
